codex-rs/core/src/config/mod.rs

```diff
@@ -69,6 +71,147 @@ pub(crate) const PROJECT_DOC_MAX_BYTES: usize = 32 * 1024; // 32 KiB
 
 pub const CONFIG_TOML_FILE: &str = "config.toml";
 
+const ALL_APPROVAL_POLICIES: [AskForApproval; 4] = [
+    AskForApproval::UnlessTrusted,
+    AskForApproval::OnFailure,
+    AskForApproval::OnRequest,
+    AskForApproval::Never,
+];
+
+#[derive(Deserialize, Debug, Clone, Default, PartialEq)]
+pub struct AdminPolicyToml {
+    pub allowed_approval_policies: Option<Vec<AskForApproval>>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct AdminPolicy {
+    pub allowed_approval_policies: Vec<AskForApproval>,
+}
+
+impl From<AdminPolicyToml> for AdminPolicy {
+    fn from(toml: AdminPolicyToml) -> Self {
+        Self {
+            allowed_approval_policies: toml
+                .allowed_approval_policies
+                .unwrap_or_else(|| ALL_APPROVAL_POLICIES.to_vec()),
+        }
+    }
+}
+
+#[derive(Debug, Error, PartialEq, Eq)]
+#[error("{message}")]
+pub struct ConstraintError {
+    pub message: String,
+}
+
+impl ConstraintError {
+    pub fn invalid_value(candidate: impl Into<String>, allowed: impl Into<String>) -> Self {
+        Self {
+            message: format!(
+                "value `{}` is not in the allowed set {}",
+                candidate.into(),
+                allowed.into()
+            ),
+        }
+    }
+}
+
+pub type ConstraintResult<T> = Result<T, ConstraintError>;
+
+impl From<ConstraintError> for std::io::Error {
+    fn from(err: ConstraintError) -> Self {
+        std::io::Error::new(std::io::ErrorKind::InvalidInput, err.to_string())
+    }
+}
+
+type ConstraintValidator<T> = dyn Fn(&T) -> ConstraintResult<()> + Send + Sync;
+
+#[derive(Clone)]
+pub struct Constrained<T> {
+    value: T,
+    validator: Arc<ConstraintValidator<T>>,
+}
+
+impl<T: Send + Sync> Constrained<T> {
+    pub fn new(
+        initial_value: T,
+        validator: impl Fn(&T) -> ConstraintResult<()> + Send + Sync + 'static,
+    ) -> ConstraintResult<Self> {
+        let validator: Arc<ConstraintValidator<T>> = Arc::new(validator);
+        validator(&initial_value)?;
+        Ok(Self {
+            value: initial_value,
+            validator,
+        })
+    }
+
+    pub fn allow_any(initial_value: T) -> Self {
```

Comment:
This doesn't help the UI know what the supported values are, though? I guess that is a limitation of `ConstraintValidator` in general?